Rollover 2.0

da-kami opened the pull request:
This implements a more resilient rollover protocol where the taker proposes a `commit_txid` from which to roll over from to the maker. Protocol details can be found here: https://github.com/itchysats/itchysats/issues/1775

TODO:

- [x] Assert on the exact fees at the end of the rollover tests, make sure that the fees did not change compared to the code on master.
- [x] Add actor tests where we rollover twice and on the second rollover the taker uses the same `event_id`. (This might be a bit tricky)

Resolves [#1622](https://github.com/itchysats/itchysats/issues/1622), https://github.com/itchysats/itchysats/issues/1775

--- 

Implications when merging this PR and rolling it into production:

Previous ItchySats version that included the `libp2p` `rollover` protocol will not be compatible because the protocol contains breaking changes. Since we did *not* roll any version that included `libp2p` `rollover` out on Umbrel there should not be any users affected. 
The legacy rollover should not be affected by the breaking changes. The feature was implemented in a way that is backwards compatible (to the legacy version). This has to be tested on `testnet`.

da-kami committed:
Remove `Msg3` from protocol

In the new design we don't need an `ack` message at the end.
See https://github.com/itchysats/itchysats/issues/1775#issuecomment-1160081024 for details

da-kami committed:
Save `settlement_event_id` with `RevokeCommit`

This is necessary so the maker can make decisions if a rollover can be started from a specific commit.
The maker has to validate that the event-id associated with the commit is not in the past yet, otherwise a rollover is not allowed anymore.

For backwards compatibility the `settlement_event_id` in `revoke_commit_transactions` is an optional field.
One all legacy cfds are closed we could add a `NOT NULL` constraint and remove the `Option` from the business logic.

da-kami committed:
Use `CompleteFee` for rollover

Instead of accumulating the fees we apply the complete fee at the end of each rollover in `RolloverCompleted`.
This is necessary, because the maker now allows multiple rollovers for the same `event_id` resulting in problems when accumulating the `funding_fee`s when applying `RolloverCompleted` events.
By saving the complete fee we introduce a snapshot at that point in time.

To ensure backwards compatibility we make the `complete_fee` in `RolloverCompleted` is optional.
If present we create the `FeeAccount` from the `complete_fee` otherwise we use the old logic of accumulating using the `funding_fee` of the rollover.

Additionally, we introduce the `CompleteFee` in the `RevokeCommit` to provide a fee-snapshot for previous rollovers.
This enables the maker to accept rollover requests for a previous `commit_txid`.
To ensure backwards compatibility we make the `complete_fee` in `RevokeCommit` is optional.

bonomat commented on daemon/src/projection.rs:
```diff
@@ -418,12 +418,22 @@ impl Cfd {
+                } else {
+                    self.aggregated.fee_account =
+                        self.aggregated.fee_account.add_funding_fee(funding_fee);
+                }
```
Is clippy not complaining about this? I think it would be nicer to use a `match` statement.

bonomat commented on daemon/src/rollover/taker.rs:
```diff
@@ -270,8 +272,13 @@ impl Actor {
 
-                            let revoked_commit =
-                                finalize_revoked_commits(&dlc, own_cfd_txs.commit.1, msg2)?;
+                            // TODO: Shouldn't the adaptor sig be the one of the dlc.commit_tx?
```
You might be right: we want to store the _revoked_ commit signatures, i.e. the one from the last commit transaction which should be in `dlc.commit`. @luckysori : what do you think ? 
Was this is a bug?

da-kami replied:
I added this TODO on the way when I noticed this, thanks for reacting! I will move this into a separate PR so we can discuss it in context!

da-kami replied:
see https://github.com/itchysats/itchysats/pull/2280

da-kami committed:
Maker allows rollover from `commit_txid`

The taker sends a `commit_txid` from where to roll over in the rollover `Proposal`.
Based on this `commit_txid` the maker picks the `event_id` from which to roll over. The costs are determined according to that `event_id`.
The maker does not allow to roll over from event id that has likely occurred.

This allows us to retry rollovers as the taker.
If the maker is ahead by one rollover, the taker can trigger a rollover with the previous event id and the maker will roll over from that point.

da-kami committed:
Maker fires and forgets last message

In the new protocol the maker ensures to finish the protocol and append a `RolloverCompleted` if the maker received the revocation keys from the taker.
If the maker fails to send the revocation keys to the taker (or the taker fails to receive) the expectation is that the taker will fail the rollover and retry using the same `event-id` as before.

This is safe for the maker because it ensures to complete (the taker cannot attack the maker by closing the connection).
This is resilient for maker and taker because the taker can retry there is a failure.

bonomat commented on sqlite-db/migrations/20220622055305_add_complete_fee_to_rollover_tables.sql:
```diff
@@ -0,0 +1,19 @@
+    rollover_completed_event_data
+ADD
+    -- The complete fee including this rollover (contrary to the funding_fee this field contains the accumulated fees until now).
+    COLUMN complete_fee text NULL;
```
I think this should be `INTEGER`

bonomat commented on sqlite-db/migrations/20220622055305_add_complete_fee_to_rollover_tables.sql:
```diff
@@ -0,0 +1,19 @@
+    -- The complete fee including this rollover (contrary to the funding_fee this field contains the accumulated fees until now).
+    COLUMN complete_fee text NULL;
+ALTER TABLE
+    rollover_completed_event_data
```
You could also combine these two alter statements: 

```
ALTER TABLE
    rollover_completed_event_data
ADD
    COLUMN complete_fee text NULL,
    COLUMN complete_fee_flow text NULL;
```

da-kami replied:
I tried, but could not get it working. I think this it unfortunately not possible in `sqlite`, see: 

https://stackoverflow.com/questions/6172815/sqlite-alter-table-add-multiple-columns-in-a-single-statement
https://www.sqlite.org/lang_altertable.html

bonomat replied:
oh, learned something new today. Thanks for links.

bonomat commented on sqlite-db/src/closed.rs:
```diff
@@ -297,8 +297,16 @@ impl ClosedCfdInputAggregate {
+                funding_fee,
+                complete_fee,
+            } => {
+                if let Some(complete_fee) = complete_fee {
```
same as earlier, I think it would be nicer to have match statements

bonomat commented on sqlite-db/src/models.rs:
```diff
@@ -901,3 +901,138 @@ impl From<Settlement> for model::Settlement {
+    None,
+}
+
+pub fn into_complete_fee(
```
Maybe this can be changed to a `impl From` block?

```
impl From<(Option<FeeFlow>, Option<i64>) for Option<model::CompleteFee> {

}
```

da-kami replied:
After giving this a quick try: Not possible because `Option` is not defined within the crate so we run into:

```
Only traits defined in the current crate can be implemented for arbitrary types [E0117]
```

I don't think there is a way around this at the moment - if you know one let me know :)

bonomat replied:
oh well, you could probably simplify it to

```
impl From<(FeeFlow, i64) for model::CompleteFee {

}
```

but feel free to leave as is.

da-kami replied:
I'd have to deal with the `Options` multiple times upon conversion then, because this code is re-used by two tables.
The `Options` are actually quite annoying - and only needed to be backwards compatible. I think I'll leave it for now and focus on the tests

bonomat commented on daemon/src/rollover/protocol.rs:
```diff
@@ -232,22 +231,22 @@ pub enum CompleteFee {
+impl From<model::CompleteFee> for CompleteFee {
+    fn from(complete_fee: model::CompleteFee) -> Self {
+        match complete_fee {
+            model::CompleteFee::LongPaysShort(a) => CompleteFee::LongPaysShort(a),
```
I'm dreaming of a world where we get rid of `LongPaysShort` and `ShortPaysLong` terminology and just express everything with positive and negative numbers...

da-kami replied:
I am not sure that will make the code easier to read and understand, but feel free to refactor this at some point :)

klochowicz replied:
We had that in the beginning, didn't we? then we decided that it will be easier and less confusing to use this terminology. Happy to move back to positive and negative numbers, as long as we have a clear doc where we state explicitly what is what.

da-kami replied:
> We had that in the beginning

IMHO this is only half of the truth - what we had initially was incomplete. To make it complete we introduced `FeeAccount` and other structs on the way. I acknowledge that changing the logic of`+` and `-` instead to an `enum` might mean more code, but I would be careful stating that it can easily be made "better" by using `+` and `-`. We re-wrote this code three times already. 

 I am happy to see another refactoring to this part as long as it has at least the same test coverage that we have now :)

bonomat commented on model/src/cfd.rs:
```diff
@@ -898,10 +898,64 @@ impl Cfd {
+                Some(revoke_commit) => revoke_commit.settlement_event_id.ok_or_else(|| anyhow!(
+                    "Proposed commit-txid {} not eligible for rollover because no event-id attached", from_tx_id
+                ))?,
+                None => bail!("Unknown commit-txid {} proposed by taker", from_tx_id),
```
This error would mean the taker has a commit-txid the maker doesn't know about right? Meaning, the taker succeeded in a former rollover but the maker failed?

da-kami replied:
Yes. Given that the maker now fires and forgets the last message this should never happen, unless the `maker` fails on the "last mile", e.g. when saving `RolloverCompleted` into the database.
I don't think we can overcome that problem without the maker messaging the taker about a failure. 
The taker could then roll-over from the previous event, but I am not sure about the security assumptions for such a scenario. Definitely out of scope for now.

bonomat commented on model/src/cfd.rs:
```diff
@@ -913,14 +967,25 @@ impl Cfd {
+                .as_ref()
+                .context("Cannot roll over without DLC")?
+                .settlement_event_id
+        };
```
I think this would benefit from a short comment why we are taking `self.dlc.settlement_event_id`.

da-kami replied:
This is what was previously picked (inside the function `hours_to_extend_in_rollover_based_on_event`) but I'm happy to add a comment!

bonomat commented on model/src/cfd.rs:
```diff
@@ -913,14 +967,25 @@ impl Cfd {
-                self.hours_to_extend_in_rollover_based_on_event(candidate_event_id, now)?
-            }
+            rollover::Version::V3 => self.hours_to_extend_in_rollover_based_on_event(
+                candidate_event_id,
```
Below you rename the arg to `to_event_id`. I think it might be worth also renaming it here.

bonomat commented on model/src/cfd.rs:
```diff
@@ -913,14 +967,25 @@ impl Cfd {
+
+        let from_event_id = if let Some(from_event_id) = from_event_id {
+            from_event_id
+        } else {
```
What do you think of adding a log message saying that this is a retry? 
That way we can track how often a rollover is being retried.

da-kami replied:
Note: I added the statement when we `start` the rollover because I think it makes more sense to log the information there. 
At this point in time we cannot say for sure if it is a retry, because the `start_rollover_maker` logic sets the `from_event_id` also when it is the current one. I implemented it this way because I think it makes the feature more complete. The fact that we have to evaluate if `from_event_id` is `Some` in `accept_rollover_proposal` is just due to being backwards compatible. 

In the future we can opt for making the `settlement_event_id` always set in `RevokeCommit`, i.e. remove the `Option` and then this check for `Some` in `accept_rollover_proposal` would go away.

bonomat reviewed (approved):
Good job. A few minor comments but overall concept should be sound

da-kami committed:
Add retry rollover test

da-kami commented on daemon/src/rollover/maker.rs:
```diff
@@ -120,28 +122,32 @@ impl Actor {
-            tracing::warn!(%order_id, "Rollover failed after handling taker proposal: {e:#}");
+            Ok(event_id) => event_id,
+            Err(e) => {
+                tracing::warn!(%order_id, "Rollover failed after handling taker proposal: {e:#}");
```
Note: I think we should change this into a `reject` scenario. The maker can send a `reject` back to the taker at this point. 

I did not change this yet because I am not sure about side-effects (I think they should be minimal, but still).

da-kami replied:
see https://github.com/itchysats/itchysats/issues/2294

klochowicz commented on daemon/src/rollover/taker.rs:
```diff
@@ -51,6 +53,8 @@ impl xtra::Actor for Actor {
     pub maker_peer_id: PeerId,
+    pub from_commit_txid: Txid,
+    pub from_settlement_event_id: BitMexPriceEventId,
 }
```
why not be explicit and also state `until_settlement_event_id` here as well? Then it would be unambiguous until when we could extend the rollover?
I believed that what #1622 originally imagined too.

da-kami replied:
The maker already tells the taker what event to use. I am not sure what value it would have to add it to the taker's proposal.

klochowicz commented on model/src/cfd.rs:
```diff
@@ -780,7 +784,10 @@ impl Cfd {
+    pub fn can_auto_rollover_taker(
+        &self,
+        now: OffsetDateTime,
+    ) -> Result<(Txid, BitMexPriceEventId), NoRolloverReason> {
```
it would be great to introduce some type that discribes this tuple, because it's pretty random in this context.

for example:
```rust
struct RolloverStartParams {
from_commit_tx: Txid,
from_settlement_event_id: BitMexPriceEventId
}
```

klochowicz commented on model/src/cfd.rs:
```diff
@@ -894,10 +903,74 @@ impl Cfd {
         self,
         tx_fee_rate: TxFeeRate,
         funding_rate: FundingRate,
+        from_params: Option<(BitMexPriceEventId, CompleteFee)>,
```
again, it's unfortunate to send `None`. Rust can be really descriptive, it would be nicer to send an enum that clarifies this better :)

da-kami replied:
Note that all these are only `Option`s because of backwards compatibility for the legacy protocol versions. 

I would hope to clean that up once we are a few versions ahead. If you feel strong about it feel free to do a small refactoring round as a follow up!

klochowicz commented on model/src/cfd.rs:
```diff
@@ -1538,15 +1643,15 @@ impl Cfd {
             bail!(
-                "Cannot rollover if candidate event ID is no later than current event ID:
-                 {candidate_settlement_time} <= {current_settlement_time}",
+                "Cannot rollover if to event ID is not later than from event ID:
```
nit: "to" and "from" don't read too well. perhaps "until" instead of "to" would make it easier? 

or maybe "start" and "end" as a pair?

klochowicz commented on model/src/cfd.rs:
```diff
@@ -2216,6 +2331,10 @@ pub struct RevokedCommit {
+    // To enable the taker to rollover from a previous commit-tx
+    // The maker requires these parameters to reason about the previous commit
+    pub settlement_event_id: Option<BitMexPriceEventId>,
+    pub complete_fee: Option<CompleteFee>,
```
this comment has a high likelihood to become stale (maybe we could extract a struct where we could put these 2 fields, so that it will never get out of place?)

da-kami replied:
fair enough, I changed it into more descriptive doc comments

klochowicz commented on model/src/cfd.rs:
```diff
@@ -3345,15 +3465,15 @@ mod tests {
     #[test]
     fn given_current_settlement_in_12_hours_and_candidate_in_19_then_7_hour_extension() {
         for now in common_time_boundaries() {
-            let current_event_id = BitMexPriceEventId::with_20_digits(now + 12.hours());
-            let candidate_event_id = BitMexPriceEventId::with_20_digits(now + 19.hours());
```
actually, "current" and "candidate" naming sounds pretty good too. I wouldn't mind having this one throughout.

da-kami replied:
Yeah, I am not super happy with `to` and `from` neither, but I found `candidate` super confusing, because for me it indicates that it is only a candidate but might not be used (i.e. something else would be used instead). That's not what we want to express here. It's literally a to-from scenario (from a cost perspective, as in the hours from and to you get charged for). Happy to see a follow up on that that changes it throughout (but please not candidate :)

klochowicz reviewed (approved):
I have added some comments, but nothing blocking; happy to fix most of the nitpicks myself in a follow-up. 

concept ACK.
